### PR TITLE
add user model

### PR DESCRIPTION
**Pull request by shiyi001**
First add the user model as an example.

### CODE REVIEW

**Comment by shiyi001** [comment]
It seems that the current code does not pass CI. @Time1ess

**Comment by Time1ess** [comment]
Please refer to the CI details about why your PR fails to build.

https://travis-ci.com/github/SPS-2020-Team-15/Graminsta/builds/177285627

**Comment by shiyi001** [comment]
I only added the most basic attributes of the User class. What else do you think can be added? @isabella0428 @hht-hannah @ArianaGu

**Comment by isabella0428** [comment]
> I only added the most basic attributes of the User class. What else do you think can be added? @isabella0428 @hht-hannah @ArianaGu

The user ids that the user follows?
And also we talk about timeline during the meeting on Monday. Maybe also maintain a list of post ids of the user's timeline?

**Comment by Time1ess** [comment]
Personally, I prefer extending the existing User model option, instead of defining our own, otherwise we need to take care a lot of details, like password hashing. PTAL https://docs.djangoproject.com/en/3.0/topics/auth/customizing/#extending-the-existing-user-model

**Comment by Time1ess** [comment]
To anwser @isabella0428 's question, we can start with a MVP one, and add fields one by one. But yes, we do need some
```
   following_users = models.ManyToManyField(..., reverse_name = 'xxx')
```

later

**Comment by Time1ess** [comment]
To be more specific, we can have one `UserInfo` model

```
class UserInfo(models.Model):
  user = models.OneToOneField(User, xxx)
  age = ...
  gender = ...
  following_users = ...
  ...
```

**Comment by shiyi001** [comment]
> To be more specific, we can have one `UserInfo` model
> 
> ```
> class UserInfo(models.Model):
>   user = models.OneToOneField(User, xxx)
>   age = ...
>   gender = ...
>   following_users = ...
>   ...
> ```

I have rewritten the code as suggested.

File: backend/graminsta/user/models.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+
+# Create your models here.
+class UserInfo(models.Model):
+    """UserInfo Class
+    """
+    user = models.OneToOneField(User, on_delete=models.CASCADE)
+    age = models.IntegerField()
+    gender = models.CharField(max_length=10)
+    reg_time = models.DateTimeField(auto_now=True)
```
Comment (Time1ess) [inline backend/graminsta/user/models.py]: I think Django user model already have this field.

See also: https://docs.djangoproject.com/en/3.0/ref/contrib/auth/#django.contrib.auth.models.User.date_joined

File: backend/graminsta/user/models.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    following_users = models.ManyToManyField(
+        "self",
+        symmetrical=False,
+        related_name="following_users"
```
Comment (Time1ess) [inline backend/graminsta/user/models.py]: This is used to do reverse look up.

For example, suppose `related_name="followed_by"', we have user A, B, C.

* User B is following User A
* User C is following User A
* User C is following User B

So for User C: `user_info_for_user_c.following_users == [User A, User B]`
for User A: `user_info_for_user_a.followed_by == [User B, User C]`

See also: https://docs.djangoproject.com/en/3.0/ref/models/fields/#django.db.models.ForeignKey.related_name

Comment (Time1ess) [reply backend/graminsta/user/models.py]: And the follow relationship should be defined on the `User` model, so I'd suggest to define that explicitly.

See also: https://stackoverflow.com/questions/40069192/django-models-database-design-for-user-and-follower

**Review by Time1ess — CHANGES_REQUESTED** [review]
I'd suggest to name this app as `core` or `auth` instead of `user`.

File: backend/graminsta/user/models.py (hunk 2 above)
Comment (Time1ess) [reply backend/graminsta/user/models.py]: And it looks like the follow/followed by relationship is tightly coupled with the `Post` app, so maybe define the relationship in that app would be more reasonable.

**Review by Time1ess — CHANGES_REQUESTED** [review]

**Comment by shiyi001** [comment]
- Change the name of app into `core` 
- Remove the follow/unfollow relationship.

**Review by Time1ess — CHANGES_REQUESTED** [review]
Please generate migration files and submit them too, otherwise people will generate migration files on their own, which might make the migration diverge.

File: backend/graminsta/core/models.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+from django.contrib.auth import models as auth_models
+
+
+# Create your models here.
+class UserInfo(models.Model):
+    """UserInfo Class
+    """
+    user = models.OneToOneField(auth_models.User, on_delete=models.CASCADE)
+    age = models.IntegerField()
+    gender = models.CharField(max_length=10)
```
Comment (Time1ess) [inline backend/graminsta/core/models.py]: For enumerable fields, use `choices` instead.

See also: https://code.djangoproject.com/ticket/18062

File: backend/graminsta/core/models.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+from django.db import models
+from django.contrib.auth import models as auth_models
+
+
+# Create your models here.
+class UserInfo(models.Model):
+    """UserInfo Class
+    """
+    user = models.OneToOneField(auth_models.User, on_delete=models.CASCADE)
+    age = models.IntegerField()
```
Comment (Time1ess) [inline backend/graminsta/core/models.py]: Nit: `PositiveIntegerField`?

**Review by Time1ess — CHANGES_REQUESTED** [review]
You might need to delete the migration files for the new changes and re-generate them.

File: backend/graminsta/core/models.py [hunk 5]
```diff
@@ -12,5 +12,9 @@ class UserInfo(models.Model):
     """UserInfo Class
     """
     user = models.OneToOneField(auth_models.User, on_delete=models.CASCADE)
-    age = models.IntegerField()
-    gender = models.CharField(max_length=10)
+    age = models.PositiveIntegerField()
+
+    GENDER_MALE = 0
+    GENDER_FEMALE = 1
+    GENDER_CHOICES = [(GENDER_MALE, 'Male'), (GENDER_FEMALE, 'Female')]
```
Comment (Time1ess) [inline backend/graminsta/core/models.py]: Move the definitions of the genders above, before all the `models.xxx` declarations. 

Nit: use Tuple (`GENDER_CHOICES = ((GENDER_MALE, ...))`) for `GENDER_CHOICES`.

**Review by Time1ess — DISMISSED** [review]

**Review by isabella0428 — APPROVED** [review]

**Review by ArianaGu — APPROVED** [review]

**Pull request merged**